### ISSUE TITLE
Unexpected performance result from Decimal vs. Nullable<Decimal>

### ISSUE THREAD

**Issue by joshudson** [dotnet/runtime#55239]
This isn't a serious problem right now; can somebody tag this discussion.

### Description

We discovered while attempting to remove some boxing that `Nullable<Decimal>` was much slower than` Decimal`, to the point where it's nearly as expensive as the boxing itself in some cases. Anyway, this reduced to a really small performance benchmark showing about 2x speed difference.

https://share.linqpad.net/5efkfl.linq

### Configuration

.NET 5: dotnet --version reports 5.0.301
OS: Windows x64

We got a slightly different results between AMD and Intel CPUs, but since we're comparing relative speeds the actual processor speed shouldn't matter much. My own clock speed is 3.84GHz on the AMD.

### Regression?

No?

### Data

````
<Query Kind="Statements" />

const int Size = 100_000_000;

{
    var sw = Stopwatch.StartNew();
    decimal total = 0;
    for (int i = 0; i < Size; i++) {
        decimal b = i;
        total += b;
    }
    Console.WriteLine("Plain total: {0} Elapsed: {1}", total, sw.Elapsed);
}

{
    var sw = Stopwatch.StartNew();
    decimal? total = 0;
    for (int i = 0; i < Size; i++) {
        decimal? b = i;
        total += b;
    }
    Console.WriteLine("Nullable total: {0} Elapsed: {1}", total, sw.Elapsed);
}


decimal HandrolledNullableAdd(decimal a, decimal b) {
    if (a == decimal.MinValue || b == decimal.MinValue) return decimal.MinValue;
    return a + b;
}

{
    var sw = Stopwatch.StartNew();
    decimal total = 0;
    for (int i = 0; i < Size; i++) {
        decimal b = i;
        total = HandrolledNullableAdd(total, b);
    }
    Console.WriteLine("Handrolled total: {0} Elapsed: {1}", total, sw.Elapsed);
}
````

````
Plain total: 4999999950000000 Elapsed: 00:00:00.5202703
Nullable total: 4999999950000000 Elapsed: 00:00:02.2274373
Handrolled total: 4999999950000000 Elapsed: 00:00:02.0125910
````

### Analysis

We don't have a working hypothesis. Small-changes to the handrolled replacement make for big changes in the result. Maybe tagging it with [MethodImpl(MethodImplOptions.AggressiveInlining)] would make for a more stable benchmark. But that's besides the point.

**Comment by huoyaoyuan** [comment]
> where it's nearly as expensive as the boxing itself in some cases.

There's no boxing in `Nullable`. It's mostly branching.

Your handrolled code does simulate `Nullable<Decimal>`, however `Nullable<Decimal>` is larger in size, thus results in slightly worse performance.

**Comment by huoyaoyuan** [comment]
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGATGGACW+bABsGAWXIAKAcNETsASgYBeAHwNsDANQNyAbi496DeSPEB+KaTmDLYmyvVad+ozQC+QA

The asm shows that the most difference is moving values around the stack and registers. Maybe there's some space to optimize though.

**Comment by joshudson** [comment]
> There's no boxing in `Nullable`. It's mostly branching.

Correct. We discovered this while removing some boxed decimals where nulls could happen so we couldn't easily change directly to `decimal`.
